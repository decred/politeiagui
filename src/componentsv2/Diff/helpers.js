const markAsAdded = elem => ({
  value: elem.value,
  lineIndex: elem.index,
  removed: false,
  added: true
});

const markAsRemoved = elem => ({
  lineIndex: elem.index,
  removed: elem.value,
  added: false
});

const markAsUnchanged = elem => ({
  value: elem.value,
  lineIndex: elem.index,
  removed: false,
  added: false
});

export const arrayDiff = (newCommentBody, oldCommentBody, lineDiffFunc) => [
  ...newCommentBody.filter(lineDiffFunc(oldCommentBody)).map(markAsAdded),
  ...oldCommentBody.filter(lineDiffFunc(newCommentBody)).map(markAsRemoved),
  ...newCommentBody.filter(lineEqFunc(oldCommentBody)).map(markAsUnchanged)
];

export const lineDiffFunc = arr => elem =>
  !arr.some(arrelem => arrelem.value === elem.value);

export const lineEqFunc = arr => elem => !lineDiffFunc(arr)(elem);

export const getLineArray = string =>
  string && string.length
    ? string.split("\n").map((line, index) => ({ value: line, index: index }))
    : [];
