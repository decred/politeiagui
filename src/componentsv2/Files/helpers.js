export function getFormattedFiles({ base64, fileList }) {
  return Array.from(fileList).map(({ name, size, type: mime }, idx) => {
    return {
      name,
      mime: name.includes(".txt") ? `${mime}; charset=utf-8` : mime,
      size,
      payload: base64[idx].split("base64,").pop()
    };
  });
}
