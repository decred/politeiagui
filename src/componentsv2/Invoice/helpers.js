import {
  INVOICE_STATUS_NEW,
  INVOICE_STATUS_UPDATED,
  INVOICE_STATUS_DISPUTED,
  INVOICE_STATUS_REJECTED,
  INVOICE_STATUS_APPROVED,
  INVOICE_STATUS_PAID
} from "src/containers/Invoice";

export const getInvoiceStatusTagProps = (invoice) => {
  switch (invoice.status) {
    case INVOICE_STATUS_NEW:
    case INVOICE_STATUS_UPDATED:
      return {
        type: "blackTime",
        text: "Waiting for review"
      };
    case INVOICE_STATUS_DISPUTED:
      return { type: "bluePending", text: "On dispute" };
    case INVOICE_STATUS_REJECTED:
      return { type: "grayNegative", text: "Rejected" };
    case INVOICE_STATUS_APPROVED:
      return { type: "yellowTime", text: "Waiting payment" };
    case INVOICE_STATUS_PAID:
      return { type: "greenCheck", text: "Paid" };
    default:
      break;
  }
};
