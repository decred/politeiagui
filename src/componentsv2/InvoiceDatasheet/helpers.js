import {
  fromHoursToMinutes,
  fromUSDCentsToUSDUnits,
  fromMinutesToHours,
  fromUSDUnitsToUSDCents
} from "../../helpers";
import { ModalEditorWrapper, selectWrapper } from "./wrappers";

import styles from "./InvoiceDatasheet.module.css";

export const columnTypes = {
  TYPE_COL: 1,
  DOMAIN_COL: 2,
  SUBDOMAIN_COL: 3,
  DESC_COL: 4,
  PROP_TOKEN_COL: 5,
  LABOR_COL: 6,
  EXP_COL: 7,
  SUBTOTAL_COL: 8
};

export const { LABOR_COL, EXP_COL, SUBTOTAL_COL } = columnTypes;

const domainOptions = [
  "Development",
  "Marketing",
  "Design",
  "Research",
  "Documentation",
  "Community Management"
];

const getDomainOptions = () =>
  domainOptions.map((op) => ({
    value: op,
    label: op
  }));

const getTypeOptions = () => [
  {
    label: "Labor",
    value: 1
  },
  {
    label: "Expense",
    value: 2
  },
  {
    label: "Misc",
    value: 3
  }
];

const defaultDomainOption = getDomainOptions()[0].value;

const defaultTypeOption = getTypeOptions()[0].value;

export const generateBlankLineItem = () => ({
  type: defaultTypeOption,
  domain: defaultDomainOption,
  subdomain: "",
  description: "",
  proposaltoken: "",
  labor: 0,
  expenses: "",
  subtotal: ""
});

export const convertLineItemsToGrid = (
  lineItems,
  readOnly = true,
  errors,
  userRate = 0,
  proposalsTokens
) => {
  const grid = [];
  const { grid: gridBody, expenseTotal, laborTotal, total } = lineItems.reduce(
    (acc, line, idx) => {
      const isLabelReadonly =
        line.type === 2 ? true : line.type === 3 ? true : readOnly;
      const isExpenseReadonly = line.type === 1 ? true : readOnly;
      const laborHours = +fromMinutesToHours(line.labor);
      const expenses = +fromUSDCentsToUSDUnits(line.expenses);
      const lineSubTotal = laborHours * userRate + expenses;
      const rowErrors = (errors && errors[idx]) || {};
      const newLine = [
        { readOnly: true, value: idx + 1 },
        {
          readOnly,
          value: line.type,
          error: rowErrors.type,
          dataEditor: selectWrapper(getTypeOptions())
        },
        {
          readOnly,
          value: line.domain,
          error: false,
          dataEditor: selectWrapper(getDomainOptions())
        },
        {
          readOnly,
          value: line.subdomain,
          error: rowErrors.subdomain
        },
        {
          readOnly,
          value: line.description,
          dataEditor: ModalEditorWrapper,
          error: rowErrors.description,
          className: styles.multilineCellValue
        },
        {
          readOnly,
          value: line.proposaltoken,
          error: rowErrors && rowErrors.proposaltoken,
          dataEditor: selectWrapper(
            proposalsTokens.map((token) => {
              return { label: token, value: token };
            })
          )
        },
        {
          readOnly: isLabelReadonly,
          value: laborHours,
          error: rowErrors && rowErrors.labor
        },
        {
          readOnly: isExpenseReadonly,
          value: expenses,
          error: rowErrors.expenses
        },
        {
          readOnly: true,
          value: lineSubTotal
        }
      ];
      return {
        grid: acc.grid.concat([newLine]),
        expenseTotal: acc.expenseTotal + line.expenses,
        laborTotal: acc.laborTotal + line.labor,
        total: acc.total + lineSubTotal
      };
    },
    { grid: [], expenseTotal: 0, laborTotal: 0, total: 0 }
  );
  const totalsLine = [
    { readOnly: true },
    { readOnly: true },
    { readOnly: true },
    { readOnly: true },
    { readOnly: true },
    {
      readOnly: true,
      value: "Total",
      forceComponent: true
    },
    {
      readOnly: true,
      value: +fromMinutesToHours(laborTotal)
    },
    {
      readOnly: true,
      value: +fromUSDCentsToUSDUnits(expenseTotal)
    },
    {
      readOnly: true,
      value: total
    }
  ];
  return grid.concat(gridBody).concat([totalsLine]);
};

export const convertGridToLineItems = (grid) => {
  const copyGrid = grid.map((row) => [...row]);
  return copyGrid.reduce((acc, rowValues, row) => {
    // skip last row
    if (row === copyGrid.length - 1) return acc;

    const lineItem = rowValues.reduce((acc, cell, col) => {
      switch (col) {
        case columnTypes.TYPE_COL:
          return { ...acc, type: +cell.value };
        case columnTypes.DOMAIN_COL:
          return { ...acc, domain: cell.value };
        case columnTypes.SUBDOMAIN_COL:
          return { ...acc, subdomain: cell.value };
        case columnTypes.DESC_COL:
          return { ...acc, description: cell.value };
        case columnTypes.PROP_TOKEN_COL:
          return { ...acc, proposaltoken: cell.value };
        case columnTypes.LABOR_COL:
          return { ...acc, labor: fromHoursToMinutes(+cell.value) };
        case columnTypes.EXP_COL:
          return { ...acc, expenses: fromUSDUnitsToUSDCents(+cell.value) };
        case columnTypes.SUBTOTAL_COL:
          return { ...acc, subtotal: +cell.value };
        default:
          return acc;
      }
    }, {});

    return acc.concat([lineItem]);
  }, []);
};

export const lineitemsWithSubtotal = (lineItems, rate) =>
  lineItems.map((l) => {
    return l.type === 1
      ? {
          ...l,
          subtotal: +fromMinutesToHours(l.labor) * +fromUSDCentsToUSDUnits(rate)
        }
      : {
          ...l,
          subtotal: +fromUSDCentsToUSDUnits(l.expenses)
        };
  });

export const createTableHeaders = () => [
  { readOnly: true, value: "", width: 25 },
  { value: "Type", width: 120, readOnly: true },
  { value: "Domain", width: 175, readOnly: true },
  { value: "Subdomain", width: 175, readOnly: true },
  { value: "Description", readOnly: true },
  { value: "Proposal Token", readOnly: true },
  { value: "Labor (hours)", width: 60, readOnly: true },
  { value: "Expense (USD)", width: 60, readOnly: true },
  { value: "Subtotal (USD)", width: 60, readOnly: true }
];

export const updateGridCell = (grid, row, col, values) => {
  grid[row][col] = { ...grid[row][col], ...values };
  return grid;
};

export const processTypeColChange = (grid, { row, col, value }) => {
  // if value is 1, the expense column must be 0, so we set it to readOnly
  if (+value === 1) {
    grid = updateGridCell(grid, row, EXP_COL, { value: 0, readOnly: true });
  } else {
    // if it is not 1, we make sure it is not readOnly
    grid = updateGridCell(grid, row, EXP_COL, { readOnly: false });
  }

  // if value is 2, the labor column must be 0, so we set it to readOnly
  if (+value === 2 || +value === 3) {
    grid = updateGridCell(grid, row, LABOR_COL, { value: 0, readOnly: true });
  } else {
    // if it is not 2, we make sure it is not readOnly
    grid = updateGridCell(grid, row, LABOR_COL, { readOnly: false });
  }

  return updateGridCell(grid, row, col, { value });
};

export const processLaborColChange = (grid, { row, col, value }, userRate) => {
  // updates subtotal upon labor value entry
  grid = updateGridCell(grid, row, SUBTOTAL_COL, { value: value * userRate });
  return updateGridCell(grid, row, col, { value });
};

export const processExpenseColChange = (grid, { row, col, value }) => {
  // updates subtotal upon expense value entry
  grid = updateGridCell(grid, row, SUBTOTAL_COL, { value: value });
  return updateGridCell(grid, row, col, { value });
};

export const processSubtotalColChange = (grid, userRate) => {
  // case where user rate changes and grid must recalculate subtotal
  for (const g of grid) {
    const row = g[0].value - 1;
    const type = g[1].value;
    const value = g[6].value;
    if (type === 1) {
      grid = updateGridCell(grid, row, SUBTOTAL_COL, {
        value: value * userRate
      });
    }
  }
  return grid;
};

export const processCellsChange = (currentGrid, changes, userRate = 0) => {
  return changes.reduce(
    (acc, change) => {
      switch (change.col) {
        case columnTypes.TYPE_COL:
          acc.grid = processTypeColChange(acc.grid, change);
          return acc;
        case columnTypes.LABOR_COL:
          acc.grid = processLaborColChange(acc.grid, change, userRate);
          return acc;
        case columnTypes.EXP_COL:
          acc.grid = processExpenseColChange(acc.grid, change);
          return acc;
        case columnTypes.SUBTOTAL_COL:
          acc.grid = processSubtotalColChange(acc.grid, userRate);
          return acc;
        default:
          acc.grid = updateGridCell(acc.grid, change.row, change.col, {
            value: change.value
          });
          return acc;
      }
    },
    { grid: currentGrid }
  );
};
