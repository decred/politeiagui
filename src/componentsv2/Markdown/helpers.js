import React, { useState } from "react";
import xssFilters from "xss-filters";
import htmlParser from "react-markdown/plugins/html-parser";
import ModalExternalLink from "../ModalExternalLink";

export const htmlParserRules = htmlParser({
  isValidNode: node => {
    return node.type !== "script";
  }
});

export const traverseChildren = (el, cb) => {
  const filterChildren = c =>
    React.Children.map(c, child => traverseChildren(child, cb));
  let newElement = null;
  if (el.children) {
    newElement = {
      ...el,
      children: filterChildren(el.children)
    };
  } else if (el.props && el.props.children) {
    const filteredChildren = filterChildren(el.props.children);
    newElement = {
      ...el,
      props: {
        ...el.props,
        children: filteredChildren
      }
    };
  }
  return newElement ? cb(newElement) : cb(el);
};

export const handleFilterXss = el => {
  if (typeof el === "string") return el;
  const props = el.props;
  if (!props) {
    return el;
  }
  const newProps = {
    ...props
  };
  if (newProps.src) {
    newProps.src = xssFilters.uriInDoubleQuotedAttr(props.src);
  }
  return {
    ...el,
    props: newProps
  };
};

const isExternalLink = link => {
  // e.preventDefault();
  // Does this to prevent xss attacks
  const tmpLink = document.createElement("a");
  tmpLink.href = link;
  const externalLink =
    tmpLink.hostname && tmpLink.hostname !== window.top.location.hostname;

  return externalLink;
};

const LinkRenderer = ({ url, children }) => {
  const [modalState, setModalState] = useState(false);
  function onLinkClick(e) {
    if (isExternalLink(url)) {
      e.preventDefault();
      setModalState(true);
    }
  }
  function onCloseModal() {
    setModalState(false);
  }
  return (
    <>
      <a href={url} onClick={onLinkClick}>
        {children}
      </a>
      <ModalExternalLink show={modalState} onClose={onCloseModal} link={url} />
    </>
  );
};

const imageHandler = ({ src, alt }) => {
  return <LinkRenderer url={src}>{alt}</LinkRenderer>;
};

const linkHandler = ({ href, children }) => {
  return <LinkRenderer url={href}>{children}</LinkRenderer>;
};

const rootHandler = filterXss => el => {
  if (filterXss) {
    el = traverseChildren(el, handleFilterXss);
  }
  const { children, ...props } = el;
  return <div {...props}>{children}</div>;
};

export const customRenderers = filterXss => {
  const rootRenderer = rootHandler(filterXss);

  return {
    image: imageHandler,
    imageReference: imageHandler,
    link: linkHandler,
    linkReference: linkHandler,
    root: rootRenderer
  };
};
