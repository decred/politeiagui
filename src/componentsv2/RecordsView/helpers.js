/**
 * Filter the records from an array of tokens and object of records by token
 * @param {Array} records
 * @param {Object} recordsTokens
 * @returns {Array} records
 */
export const getRecordsByTabOption = (records, tokens) =>
  tokens.map((token) => records[token]).filter(Boolean);

/**
 * Returns the censorsihp token of a given record
 * @param {Object} record
 * @returns {String} recordToken
 */
export const getRecordToken = (record) =>
  record && record.censorshiprecord && record.censorshiprecord.token;
