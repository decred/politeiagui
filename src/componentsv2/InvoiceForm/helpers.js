import { getCurrentMonth, getCurrentYear } from "src/helpers";

/**
 * Returns the initial month and year to be applied to a blank invoice form
 * @returns {Object} { year, month}
 */
export const getInitialDateValue = () => {
  const currYear = getCurrentYear();
  const currMonth = getCurrentMonth();

  // case is december
  if (currMonth === 1) {
    return {
      year: currYear - 1,
      month: 12
    };
  }

  return {
    year: currYear,
    month: currMonth - 1
  };
};

/**
 * Returns the range of years and months allowed for selection
 * @returns {Object} { min, max }
 */
export const getMinMaxYearAndMonth = () => {
  const min = { year: 2018, month: 1 };
  return {
    min,
    max: getInitialDateValue()
  };
};
