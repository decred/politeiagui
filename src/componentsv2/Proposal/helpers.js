import distanceInWordsToNow from "date-fns/distance_in_words_to_now";
import {
  BLOCK_DURATION_MAINNET,
  BLOCK_DURATION_TESTNET,
  PROPOSAL_STATUS_PUBLIC,
  PROPOSAL_VOTING_ACTIVE,
  PROPOSAL_VOTING_AUTHORIZED,
  PROPOSAL_VOTING_NOT_AUTHORIZED,
  PROPOSAL_STATUS_ABANDONED,
  PROPOSAL_VOTING_FINISHED,
  PROPOSAL_STATUS_UNREVIEWED,
  PROPOSAL_STATUS_UNREVIEWED_CHANGES
} from "../../constants";
import { getTextFromIndexMd } from "src/helpers";

export const getMarkdownContent = files => {
  const markdownFile = files.find(f => f.name === "index.md");
  return getTextFromIndexMd(markdownFile);
};

/**
 * Returns the total amount of votes received by a given proposal
 * @param {Object} proposal
 */
export const getVotesReceived = proposal => {
  if (!proposal || !proposal.voteStatus) {
    return 0;
  }
  return proposal.voteStatus.totalvotes;
};

/**
 * Converts the api data for vote status into an array of data
 * that can be used to render the StatusBar
 * @param {Object} voteStatus
 * @returns {Array} status bar data
 */
export const getStatusBarData = voteStatus => {
  return voteStatus.optionsresult
    .map(op => ({
      label: op.option.id,
      amount: op.votesreceived,
      color: op.option.id === "yes" ? "#41BE53" : "#ED6D47"
    }))
    .sort(a => (a.label === "yes" ? -1 : 1));
};

/**
 * Return the votes quorum of a given proposal vote status
 * @param {Object} voteStatus
 */
export const getQuorumInVotes = voteStatus =>
  Math.trunc(
    (voteStatus.numofeligiblevotes * voteStatus.quorumpercentage) / 100
  );

/**
 * Returns true if the given proposal is public
 * @param {Object} proposal
 * @returns {Boolean} isPublic
 */
export const isPublicProposal = proposal => {
  return proposal.status === PROPOSAL_STATUS_PUBLIC;
};

/**
 * Returns true if the given proposal is unreviewed
 * @param {Object} proposal
 * @returns {Boolean} isUnreviewed
 */
export const isUnreviewedProposal = proposal => {
  return (
    proposal.status === PROPOSAL_STATUS_UNREVIEWED ||
    proposal.status === PROPOSAL_STATUS_UNREVIEWED_CHANGES
  );
};

/**
 * Returns true if the given proposal is public, but voting
 * is not authorized yet.
 * @param {Object} proposal
 * @returns {Boolean} isVotingNotAuthorized
 */
export const isVotingNotAuthorizedProposal = proposal => {
  return (
    !!proposal &&
    !!proposal.voteStatus &&
    proposal.voteStatus.status === PROPOSAL_VOTING_NOT_AUTHORIZED
  );
};

/**
 * Returns true if the given proposal is public, but voting
 * has finished already
 * @param {Object} proposal
 * @returns {Boolean} isVotingFinished
 */
export const isVotingFinishedProposal = proposal => {
  return (
    !!proposal &&
    !!proposal.voteStatus &&
    proposal.voteStatus.status === PROPOSAL_VOTING_FINISHED
  );
};

/**
 * Returns true if the given proposal is editable
 * @param {Object} proposal
 * @returns {Boolean} isEditable
 */
export const isEditableProposal = proposal => {
  return (
    isUnreviewedProposal(proposal) ||
    (isPublicProposal(proposal) && isVotingNotAuthorizedProposal(proposal))
  );
};

/**
 * Returns true if the given proposal is under discussion
 * @param {Object} proposal
 * @returns {Boolean} isUnderDiscussion
 */
export const isUnderDiscussionProposal = proposal => {
  return (
    isPublicProposal(proposal) &&
    !isVoteActiveProposal(proposal) &&
    !isVotingFinishedProposal(proposal)
  );
};

/**
 * Returns true if the given proposal is abandoned
 * @param {Object} proposal
 * @returns {Boolean} isAbandoned
 */
export const isAbandonedProposal = proposal => {
  return proposal.status === PROPOSAL_STATUS_ABANDONED;
};

/**
 * Returns true if the given proposal is approved
 * @param {Object} voteStatus
 * @returns {Boolean} isApproved
 */
export const isApprovedProposal = proposal => {
  if (!proposal || !proposal.voteStatus || !isPublicProposal(proposal)) {
    return false;
  }
  const {
    quorumpercentage,
    passpercentage,
    numofeligiblevotes,
    optionsresult,
    totalvotes
  } = proposal.voteStatus;
  const quorumInVotes = (quorumpercentage * numofeligiblevotes) / 100;
  const quorumPasses = totalvotes >= quorumInVotes;
  if (!quorumPasses) {
    return false;
  }

  const yesVotes = optionsresult.find(op => op.option.id === "yes")
    .votesreceived;

  return yesVotes > (passpercentage * totalvotes) / 100;
};

export const isVoteActiveProposal = proposal =>
  !!proposal &&
  !!proposal.voteStatus &&
  proposal.voteStatus.status === PROPOSAL_VOTING_ACTIVE;

/**
 * Return the properties to be passed to the StatusTag UI component
 * @param {Object} proposal
 */
export const getProposalStatusTagProps = proposal => {
  if (isPublicProposal(proposal)) {
    switch (proposal.voteStatus.status) {
      case PROPOSAL_VOTING_NOT_AUTHORIZED:
        return { type: "blackTime", text: "Hasn't authorized yet" };
      case PROPOSAL_VOTING_AUTHORIZED:
        return { type: "yellowTime", text: "Waiting for approval" };
      case PROPOSAL_VOTING_ACTIVE:
        return { type: "bluePending", text: "Active" };
      case PROPOSAL_VOTING_FINISHED:
        if (isApprovedProposal(proposal)) {
          return { type: "greenCheck", text: "Finished" };
        } else {
          return { type: "grayNegative", text: "Finished" };
        }
      default:
        break;
    }
  }

  if (isAbandonedProposal(proposal)) {
    return { type: "orangeNegativeCircled", text: "Abandoned" };
  }

  return { type: "grayNegative", text: "missing" };
};

/**
 * Return the amount of blocks left to the end of the voting period
 * @param {Object} proposal
 * @param {Number} chainHeight
 * @returns {Number} number of blocks left
 */
export const getVoteBlocksLeft = (proposal, chainHeight) => {
  if (!isPublicProposal(proposal)) return null;
  const {
    voteStatus: { endheight }
  } = proposal;
  return +endheight - chainHeight;
};

/**
 * Return a "human readable" message of how long will take until the voting ends
 * @param {Object} proposal
 * @param {Number} chainHeight
 * @param {Boolean} isTestnet
 * @returns {String} message
 */
export const getVoteTimeLeftInWords = (proposal, chainHeight, isTestnet) => {
  if (
    !proposal ||
    !isPublicProposal(proposal) ||
    proposal.voteStatus.status !== PROPOSAL_VOTING_ACTIVE
  ) {
    return "";
  }

  const blocks = getVoteBlocksLeft(proposal, chainHeight);
  const blockTimeMinutes = isTestnet
    ? blocks * BLOCK_DURATION_TESTNET
    : blocks * BLOCK_DURATION_MAINNET;
  const mili = blockTimeMinutes * 60000;
  const dateMs = new Date(mili + Date.now()); // gets time in ms

  // console.log(+endheight, chainHeight);
  return distanceInWordsToNow(dateMs, { addSuffix: true });
};
