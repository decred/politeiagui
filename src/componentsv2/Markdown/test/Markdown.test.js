import React from "react";
import ReactMarkdown from "react-markdown";
import { mount } from "enzyme";
import { customRenderers } from "../helpers";

const maliciousBodyText =
  "![clickforxss](javascript:alert('XSS'))  [clickforxss](javascript:alert('XSS'))";

it("filter potencial 'XSS' attackers", () => {
  const wrapper = mount(
    <ReactMarkdown
      source={maliciousBodyText}
      renderers={customRenderers(true)}
    />
  );
  wrapper.find("LinkRenderer").forEach(el => {
    expect(["x-javascript:alert('XSS')", "javascript:void(0)"]).toContain(
      el.prop("url")
    );
  });
});
