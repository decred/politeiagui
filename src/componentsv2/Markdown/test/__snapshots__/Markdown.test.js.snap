// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter potencial 'XSS' attackers 1`] = `
<div>
  <p>
    <a
      href="x-javascript:alert('XSS')"
      onClick={[Function]}
      rel="nofollow noopener noreferrer"
      target="_blank"
    >
      clickforxss
    </a>
      
    <a
      href="javascript:void(0)"
      onClick={[Function]}
      rel="nofollow noopener noreferrer"
      target="_blank"
    >
      clickforxss
    </a>
  </p>
</div>
`;
